### ISSUE TITLE
Is the Overlay Blending Mode Implemented Correctly?

### ISSUE THREAD

**Issue by JohnTravolski** [flrs/blend_modes#6]
It seems the the overlay blending mode produces the exact same output as the soft light blending mode, or at least it does in all of my tests.  Are you confident that it's implemented correctly?

For reference, in Photoshop and After Effects, soft light and overlay look noticeably different.

**Comment by flrs** [comment]
Hi @JohnTravolski,

No definitive answers yet for now. But there is a possibility that soft light and overlay are implemented differently in the Adobe world and in GIMP. If the tests of this package pass it means that the ,ethod implementation in blend_modes matches GIMP's methods.

Let me know if you find out more!

**Comment by JohnTravolski** [comment]
I found an interesting line here:
https://docs.gimp.org/en/gimp-concepts-layer-modes.html
"In some versions of GIMP, “Overlay” mode and “Soft light” mode are identical."
I think that may be the problem.  Are those the equations you were implementing?

If possible, it would be ideal if the version of overlay that gets implemented is different from soft light.

**Comment by JohnTravolski** [comment]
I took a look at Wikipedia's implementation of Overlay and used your code as a basis to implement is.  For example, given your code, in the overlay function definition, if I replace:
`comp = img_in[:,:,:3] * (img_in[:,:,:3] + (2 * img_layer[:,:,:3]) * (1 - img_in[:,:,:3]))`

with (using the Wikipedia formula):
```
a = img_in[:,:,:3]
b = img_layer[:,:,:3]
comp = np.less(a, 0.5)*(2*a*b) + np.greater_equal(a, 0.5)*(1 - (2*(1 - a)*(1 - b)))
```
Then the overlay blending mode gives a result that is indeed different from Soft Light and looks just like the results that After Effects gives me.

**Comment by JohnTravolski** [comment]
You mentioned that these could be sped up if implemented in Cython.  If you're interested, I have done so here, and gotten a pretty good speed improvement:
https://pastebin.com/fAD9avjP
If there're any obvious optimizations that I'm not seeing in the Cython code, feel free to let me know.  I have successfully implemented of Photoshop's blending modes, as well as yours, except the "Darker Color, Lighter Color, Hue, Saturation, Color, Luminosity" blending modes.  I attempted hue, but couldn't quite figure it out.

The only implementation difference from yours is that each blend mode function already expects the input image to be divided by 255, and the output image is not multiplied by 255.  This makes it easier to blend multiple layers together without unnecessary computation.  So you have to do those division and multiplication steps manually before and after blending.

**Comment by JohnTravolski** [comment]
Updated version that runs faster (about a 20% speedup) for 8K images (uses fused types instead of dereferencing a function pointer each time):
https://pastebin.com/N3Siqg4b

**Comment by flrs** [comment]
@JohnTravolski, thanks for doing work on the overlay code. I will implement it!

**Comment by flrs** [comment]
@JohnTravolski awesome work with the Cython implementation! I think this is something that others could profit from and we should make it part of the package.

**Comment by flrs** [comment]
Closed in favor of issues #8 and #9.

**Comment by AeroDEmi** [comment]
How can I use this Cython implementation?
